docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/Ranking/Ranking.ttinclude

```diff
@@ -0,0 +1,108 @@
+            // Setting the seed to a fixed number in this example to make outputs deterministic.
+            var mlContext = new MLContext(seed: 0);
+
+            // Create a list of training data points.
+            var dataPoints = GenerateRandomDataPoints(1000);
+
+            // Convert the list of data points to an IDataView object, which is consumable by ML.NET API.
+            var trainingData = mlContext.Data.LoadFromEnumerable(dataPoints);
+
+<# if (TrainerOptions == null) { #>
+            // Define the trainer.
+            var pipeline = mlContext.Ranking.Trainers.<#=Trainer#>();
+<# } else { #>
+            // Define trainer options.
+            var options = new <#=TrainerOptions#>;
+
+            // Define the trainer.
+            var pipeline = mlContext.Ranking.Trainers.<#=Trainer#>(options);
+<# } #>
+
+            // Train the model.
+            var model = pipeline.Fit(trainingData);
+
+            // Create testing data. Use different random seed to make it different from training data.
+            var testData = mlContext.Data.LoadFromEnumerable(GenerateRandomDataPoints(500, seed:123));
+
+            // Run the model on test data set.
+            var transformedTestData = model.Transform(testData);
+
+            // Convert IDataView object to a list.
+            var predictions = mlContext.Data.CreateEnumerable<Prediction>(transformedTestData, reuseRowObject: false).ToList();
+
+            // Look at 5 predictions
+            foreach (var p in predictions.Take(5))
+                Console.WriteLine($"Label: {p.Label}, Score: {p.Score}");
+
+            <#=ExpectedOutputPerInstance#>
+
+            // Evaluate the overall metrics
+            var metrics = mlContext.Ranking.Evaluate(transformedTestData);
+            PrintMetrics(metrics);
+            
+            <#=ExpectedOutput#>
+        }
+
+        private static IEnumerable<DataPoint> GenerateRandomDataPoints(int count, int seed = 0, int groupSize = 10)
+        {
+            var random = new Random(seed);
+            float randomFloat() => (float)random.NextDouble();
+            for (int i = 0; i < count; i++)
+            {
+                var label = random.Next(0, 5);
+                yield return new DataPoint
+                {
+                    Label = (uint)label,
+                    GroupId = (uint)(i / groupSize),
+                    // Create random features that are correlated with the label.
+                    // For data points with larger labels, the feature values are slightly increased by adding a constant.
+                    Features = Enumerable.Repeat(label, 50).Select(x => randomFloat() + x * 0.1f).ToArray()
+                };
+            }
+        }
+
+        // Example with label, groupId, and 50 feature values. A data set is a collection of such examples.
+        private class DataPoint
+        {
+            [KeyType(5)]
+            public uint Label { get; set; }
+            [KeyType(100)]            
```

Comment:
>             [](start = 26, length = 12)

can you remove this extra spaces.

Reply:
now I can see them haha :) Thank you for pointing out. 

---
In reply to: [275927787](https://github.com/dotnet/machinelearning/pull/3338#discussion_r275927787) [](ancestors = 275927787)